### DISCUSSION TITLE
[Question] How to highlight more than one instance when using set_colors

### DISCUSSION THREAD

**Discussion by CarolinaRDC** [pymupdf/PyMuPDF#1370] [Q&A] [answered]
**Configuration**

```
3.7.6 (tags/v3.7.6:43364a7ae0, Dec 19 2019, 00:42:30) [MSC v.1916 64 bit (AMD64)] 
 win32 
 
PyMuPDF 1.19.1: Python bindings for the MuPDF 1.19.0 library.
Version date: 2021-10-23 00:00:10.
Built for Python 3.7 on win32 (64-bit).
```

Hello, 
I'm trying to make a function that takes a pdf and a certain text and returns a new pdf with that text highlighted. I'm having problems with the method **update**; it doesn't seem to work properly  when it has to update more than one instance. 
However the fuction works perfectly fine with the default color, when I remove the set_colors and update methods.

The code:

```
def highlight_within_pdf(file_path, text, page_n):
    doc = fitz.open(file_path)
    page = doc.load_page(page_n)
    text_instances = page.search_for(text)

    for inst in text_instances:
        highlight = page.add_highlight_annot(inst)
        highlight.set_colors({"stroke": (0.5, 0.8, 1)})
        highlight.update()

    doc.save("highlighted_pdf.pdf")
    doc.close()
```
   

The Error:

```
Traceback (most recent call last):
  File "C:/Users/0016195/PycharmProjects/pruebas/pdf_highlight.py", line 54, in <module>
    highlight_within_pdf(file_path, text)
  File "C:/Users/0016195/PycharmProjects/pruebas/pdf_highlight.py", line 46, in highlight_within_pdf
    highlight.update()
  File "C:\Users\0016195\PycharmProjects\pruebas\venv\lib\site-packages\fitz\fitz.py", line 7555, in update
    ap_tab = ap.splitlines()  # split in single lines
AttributeError: 'NoneType' object has no attribute 'splitlines'

```

 What would it be a more suitable approach?

 
 Thank you very much in advance

**Comment by CarolinaRDC** [comment, marked as answer]
Found the solution. I only needed to load the page again after the update:

```
def highlight_within_pdf(file_path, text, page_n):
    doc = fitz.open(file_path)
    page = doc.load_page(page_n)
    text_instances = page.search_for(text)

    for inst in text_instances:
        highlight = page.add_highlight_annot(inst)
        highlight.set_colors({"stroke": (0.5, 0.8, 1)})
        highlight.update()
        page = doc.load_page(page_n)
        
    doc.save("new_pdf_highlighted.pdf", garbage=4, deflate=True, clean=True)
    doc.close
```

**Comment by JorjMcKie** [comment]
Great you figured that one out!
The issue here is that I have extended MuPDF logic somewhat. So indeed, the page needs to be "released" and then reloaded to enforce dangling updates.
1. There is a special `Document` method for this: ``page = doc.reload_page(page)`.
2. If you want you can make one single annotation out of all search hits by directly using the search rectangle / quad list as argument for adding the annotation: `page.add_highlight_annot(text_instances)` - no iteration across the single rectangles.

**Reply by CarolinaRDC** [reply]
Thank you very much for your response!. I'm definitely doing that.
